variance.json

```diff
@@ -90,5 +90,28 @@
             "href": "http://mathworld.wolfram.com/Variance.html",
             "title": "Variance explained by Wolfram MathWorld"
         }
-    ]
+    ],
+    "process_graph": {
+        "sd": {
+            "process_id": "sd",
+            "arguments": {
+                "data": {
+                    "from_parameter": "data"
+                },
+                "ignore_nodata": {
+                    "from_parameter": "ignore_nodata"
+                }
+            }
+        },
+        "power": {
+            "process_id": "power",
+            "arguments": {
+                "base": {
+                    "from_node": "sd"
+                },
+                "p": 2
+            },
+            "result": true
+        }
+    }
```

Comment:
Is it OK to have cyclic aliases (`variance` references `sd`)? Maybe we should write variance as
```
variance(data) = {
    avg = mean(data);
    return sum(apply(data, val => pow(val - avg, 2)));
}
```

Reply:
Yes, that's okay, although it's always better to have it more low-level.

Reply:
Problem is: I can't access avg in the apply callback.

Reply:
For 1 there's a PR here: #142

Reply:
I'll leave this open for discussion until the end of the day. If there's no additional motion, I'll implement option 1.

Reply:
Not sure here, I'm fine with option 1 as long as it's backwards compatible. (So everything that currently does not need context simply continues to work.)

Reply:
Process graphs and most processes are not really backward-compatible anyway, but options 1 and 2 are both "backward-compatible" in a sense that they don't break things compared to the state before merging this PR. Worst issue I see with 2 is that it get's harder to resolve dependencies in the graph and it need's an adoption of the parser, whereas 1 doesn't need a change in process graph parsing or the API at all.

Reply:
> 2. access of parent nodes in callbacks?

what do you mean with this? In the variance example, the `avg` node would not be a parent, but more some kind of sibling node of the apply callback node, right?

Reply:
> what do you mean with this? In the variance example, the avg node would not be a parent, but more some kind of sibling node of the apply callback node, right?

@soxofaan 
Looking at Mihas example:
```
    avg = mean(data);
    return sum(apply(data, val => pow(val - avg, 2)));
```
the following part is a child process graph:
`pow(val - avg, 2)`

The process graph specification only allows access to nodes in the same process graph (i.e. no access to children or parents). avg was computed in the "root" scope, but is accessed in the "child" scope and this is undefined in the "child" scope. Therefore, you either need to broaden the scope for from_node (option 2) or introduce an explicit was to pass computed data, e.g. via a context parameter (option 1). Is this better to understand?

Reply:
ok thanks, I understand, I just misread a part apparently :)

Reply:
Hey @mkadunc,

my statistics classes were long time ago, but the sum in your example must be replaced with mean, I think?!

The process graph would then be:
```json
{
        "mean": {
            "process_id": "mean",
            "arguments": {
                "data": {
                    "from_parameter": "data"
                }
            }
        },
        "apply": {
            "process_id": "apply",
            "arguments": {
                "data": {
                    "from_parameter": "data"
                },
                "process": {
                    "process-graph": {
                        "subtract": {
                            "process_id": "subtract",
                            "arguments": {
                                "x": {
                                    "from_parameter": "x"
                                },
                                "y": {
                                    "from_parameter": "context"
                                }
                            }
                        },
                        "power": {
                            "process_id": "power",
                            "arguments": {
                                "base": {
                                    "from_node": "subtract"
                                },
                                "p": 2
                            },
                            "result": true
                        }
                    }
                },
                "context": {
                    "from_node": "mean"
                }
            }
        },
        "mean2": {
            "process_id": "mean",
            "arguments": {
                "data": {
                    "from_node": "apply"
                },
                "ignore_nodata": {
                    "from_parameter": "ignore_nodata"
                }
            },
            "result": true
        }
    }
```